### PR TITLE
Sub-components support for jsx-props-no-spreading

### LINKED ISSUE #2529: react/jsx-props-no-spreading doesn't support "sub-components".
**Dependencies:**
- `"eslint": "^6.7.2"`
- `"eslint-plugin-react": "^7.17.0"`

**.eslintrc.json**
```
{
  ...
  "react/jsx-props-no-spreading": [
    "error",
    {
      "exceptions": [
        "AsyncSelect",
        "Select",
        "components.Group",
        "components.ClearIndicator",
        "components.DropdownIndicator"
      ]
    }
  ]
  ...
}
```

**What happens?**

The following patterns are considered warnings regardless they are in the exception list:

```jsx
<components.Group {...groupProps} />
<components.DropdownIndicator {...props} />
<components.ClearIndicator {...props} />
```

The following patterns are **not** considered warnings:

```jsx
<AsyncSelect {...props} />
<Select {...getCustomProps({})} />
```

### PR DESCRIPTION
**Pull request by jonathanpalma**
closes #2529

### CODE REVIEW

File: lib/rules/jsx-props-no-spreading.js [hunk 1]
```diff
@@ -31,44 +31,49 @@ module.exports = {
...
-          custom: {
-            enum: [OPTIONS.enforce, OPTIONS.ignore]
-          },
-          exceptions: {
-            type: 'array',
-            items: {
-              type: 'string',
-              uniqueItems: true
+    schema: [
+      {
```
Comment (ljharb) [inline lib/rules/jsx-props-no-spreading.js]: can you please revert the unrelated formatting changes here?

File: lib/rules/jsx-props-no-spreading.js [hunk 2]
```diff
@@ -79,22 +84,36 @@ module.exports = {
...
-          (!ignoreHtmlTags && isException(tagName, exceptions)))
+            (!ignoreHtmlTags && isException(tagName, exceptions)))
         ) {
           return;
         }
         if (
           isCustomTag &&
           ((ignoreCustomTags && !isException(tagName, exceptions)) ||
-          (!ignoreCustomTags && isException(tagName, exceptions)))
+            (!ignoreCustomTags && isException(tagName, exceptions)))
```
Comment (ljharb) [inline lib/rules/jsx-props-no-spreading.js]: here as well

File: lib/rules/jsx-props-no-spreading.js (hunk 1 above)
Comment (jonathanpalma) [reply lib/rules/jsx-props-no-spreading.js]: Done

File: lib/rules/jsx-props-no-spreading.js (hunk 2 above)
Comment (jonathanpalma) [reply lib/rules/jsx-props-no-spreading.js]: Done

File: tests/lib/rules/jsx-props-no-spreading.js [hunk 3]
```diff
@@ -27,170 +27,235 @@ const ruleTester = new RuleTester({parserOptions});
...
-      '<App>',
-      '   <img {...props}/>',
-      '   <Image src={src} alt={alt}/>',
-      '   <div {...someOtherProps}/>',
-      '</App>'
-    ].join('\n'),
-    options: [{html: 'ignore'}]
-  }, {
-    code: `
+  valid: [
```
Comment (ljharb) [inline tests/lib/rules/jsx-props-no-spreading.js]: there's some more reformatting here; can that be reverted?

Comment (jonathanpalma) [reply tests/lib/rules/jsx-props-no-spreading.js]: Done

**Pull request merged**